From fc64d32ccedccf257331eacc49f717aff1951b8c Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 28 Nov 2018 22:53:57 +0200
Subject: [PATCH 0819/1051] net: mvpp2: free non recycled buffer according to
 frag-size

If the recycle has no place to keep new recyclable buffer, it
makes free for that fragment.
The correct free-procedure depends upon fragment size vs PAGE_SIZE.
This decision should be done by frag_size
but NOT by pool-id=LONG since short JUMBO mtu (1500<MTU<=3584)
has pool-frame-size < PAGE.

Change-Id: Iafaf8427874cf839a07058e9a8b11bff2be6e326
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/61422
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index fea81b9d764a..dbe462b440ec 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3309,7 +3309,7 @@ static inline void mvpp2_recycle_put(struct mvpp2_port *port,
 		pool->pbuf[++idx] = skb;
 		pcpu->idx[MVPP2_BM_POOLS_NUM] = idx;
 		if (skb->head) {
-			if (pool_id < MVPP2_BM_JUMBO)
+			if (bm_pool->frag_size <= PAGE_SIZE)
 				skb_free_frag(skb->head);
 			else
 				kfree(skb->head);
-- 
2.17.1

